tooling/debugger/src/context.rs

```diff
@@ -142,6 +185,47 @@ impl<'a, B: BlackBoxFunctionSolver> DebugContext<'a, B> {
         }
     }
 
+    fn breakpoint_reached(&self) -> bool {
+        if let Some(location) = self.get_current_opcode_location() {
+            self.breakpoints.contains(&location)
+        } else {
+            false
+        }
+    }
+
+    pub(super) fn is_valid_location(&self, location: &OpcodeLocation) -> bool {
```

Comment:
It would be nice to have some documentation on why we can't use `self.debug_artifact.line_index(location).is_ok()` for this.

Reply:
I don't follow here. This is dealing with opcode locations, but I couldn't find any function in `debug_artifact` that takes one as parameter. Probably the function name is confusing? I'll rename to `is_valid_opcode_location` to avoid overloading the location term.

Reply:
Ah yes, I'm getting a little mixed up between `OpcodeLocation` and `Location`. I was thinking that you wanted to reject opcode locations that couldn't get mapped to source while still accepting the opcodes which get generated for circuit inputs, etc. 

The new version looks good to me.